File: src/abc/matlab/abc.m
function ABC_algorithm()
    % ABC 算法参数
    max_iter = 1000;           % 最大迭代次数
    food_number = 50;          % 蜂群大小 (假设每个蜂巢一个食物位置)
    dimension = 2;             % 问题的维度
    lower_bound = -5;          % 搜索空间的下界
    upper_bound = 5;           % 搜索空间的上界
    limit = 100;               % 限制数目（每个食物点的最大尝试次数）

    % 初始化食物位置和适应度
    food = lower_bound + (upper_bound - lower_bound) * rand(food_number, dimension);
    fitness = evaluate_fitness(food);  % 评估食物位置的适应度
    
    % 初始化蜜蜂群体
    employed_bees = food;
    onlooker_bees = food;
    
    % 迭代搜索
    for iter = 1:max_iter
        % 在雇佣蜜蜂阶段搜索新的解
        for i = 1:food_number
            new_food = search_new_food(employed_bees(i,:), lower_bound, upper_bound);
            new_fitness = evaluate_fitness(new_food);
            if new_fitness < fitness(i)
                food(i,:) = new_food;
                fitness(i) = new_fitness;
            end
        end

        % 在观察蜜蜂阶段选择食物源
        for i = 1:food_number
            prob = fitness(i) / sum(fitness);
            if rand < prob
                new_food = search_new_food(food(i,:), lower_bound, upper_bound);
                new_fitness = evaluate_fitness(new_food);
                if new_fitness < fitness(i)
                    food(i,:) = new_food;
                    fitness(i) = new_fitness;
                end
            end
        end
        
        % 更新全局最优解
        [min_fitness, best_index] = min(fitness);
        best_solution = food(best_index, :);
        
        % 输出当前最优解
        disp(['Iter: ', num2str(iter), ' Best Fitness: ', num2str(min_fitness)]);
    end
    
    % 最优解
    disp('最优解：');
    disp(best_solution);
end

% 评估适应度函数，假设目标是最小化目标函数
function fitness = evaluate_fitness(food)
    fitness = sum(food.^2, 2);  % 例如求解一个简单的二次目标函数 f(x) = x^2
end

% 生成新食物源
function new_food = search_new_food(current_food, lower_bound, upper_bound)
    phi = rand(1, numel(current_food)) * 2 - 1; % 随机步长
    step = phi .* (current_food - lower_bound);  % 调整搜索步长
    new_food = current_food + step;              % 新的食物源位置
    new_food = max(min(new_food, upper_bound), lower_bound); % 保证新位置在边界内
end

File: src/abc/c/abc.c
#include "abc.h"
#include "alg_inc.h"
#include "matrix/alg_matrix.h"
#include "memalloc/alg_memalloc.h"
#include "random/alg_random.h"
#include "vector/alg_vector.h"
#include <math.h>
#include <stdio.h>
#include <time.h>

static void abc_fresh_fitness(abc_handle *handle);

abc_handle *abc_init(int food_number, int dimension, double lower_bound, double upper_bound, int limit, evaluate_function evaluate) {
    abc_handle *handle = ALG_MALLOC(sizeof(abc_handle));
    if (handle == NULL) {
        ERROR("HANDLE INIT ERROR");
        return NULL;
    }
    handle->food_number = food_number;
    handle->lower_bound = lower_bound;
    handle->upper_bound = upper_bound;
    handle->limit = limit;
    handle->evaluate = evaluate;

    handle->food_matrix = alg_matrix_create(food_number, dimension);
    if (handle->food_matrix == NULL) {
        ERROR("INIT FOOD_MATRIX ERROR");
        ALG_FREE(handle);
        return NULL;
    }
    alg_matrix_fill_random(handle->food_matrix, lower_bound, upper_bound);

    handle->fitness = alg_vector_create(food_number, 0.0);
    if (handle->fitness == NULL) {
    }
    abc_fresh_fitness(handle);

    handle->employed_bees = alg_matrix_copy(handle->food_matrix);
    handle->onlooker_bees = alg_matrix_copy(handle->food_matrix);

    return handle;
}

static void abc_fresh_fitness(abc_handle *handle) {
    for (int i = 0; i < handle->food_number; i++) {
        alg_vector *tmp_vector = alg_vector_from_matrix_row(handle->food_matrix, i);
        alg_vector_set_val(handle->fitness, i, handle->evaluate(tmp_vector));
        alg_vector_free(tmp_vector);
    }
}
static void search_new_food(abc_handle *handle, alg_vector *current_food) {
    alg_vector *step = alg_vector_create(handle->dim, 0.0);
    if (step == NULL) {
        ERROR("PHI INIT ERROR");
        return;
    }
    for (int i = 0; i < handle->dim; i++) {
        double current_val = *alg_vector_get_val(current_food, i);
        alg_vector_set_val(step, i, alg_random_float64(handle->lower_bound - current_val, current_val - handle->lower_bound));
        alg_vector_set_val(current_food, i, *alg_vector_get_val(step, i));
    }
    alg_vector_free(step);
}

static double vector_sum(alg_vector *vec) {
    double ret = 0.0;
    for (int i = 0; i < vec->size; i++)
        ret += vec->vector[i];
    return ret;
}

double alg_safe_divide(double numerator, double denominator) {
    if (fabs(denominator) < 1e-9) {
        // 除数为零，返回 NaN (Not a Number)
        ERROR("Division by zero!");
        return NAN; // 返回 NaN，表示无效的结果
    }
    return numerator / denominator; // 正常除法
}

int abc_get_best_solution(abc_handle *handle) {
    int index = 0;
    double fit = *alg_vector_get_val(handle->fitness, 0), fit_iterval;
    for (int i = 1; i < handle->food_number; i++) {
        fit_iterval = *alg_vector_get_val(handle->fitness, i);
        if (fit > fit_iterval) {
            fit = fit_iterval;
            index = i;
        }
    }
    alg_vector *best_vec = alg_vector_from_matrix_row(handle->food_matrix, index);
    char *strs = alg_vector_print_str(best_vec);
    printf("Best solve VAL is %.2f, Best SOLVE is %s", fit, strs);
    ALG_FREE(strs);
    alg_vector_free(best_vec);
    return index;
}

alg_state abc_fresh(abc_handle *handle) {
    alg_vector *new_food;
    double sum;
    double prob;
    double rand;
    double new_fitness;

    // 1. 雇佣蜜蜂搜索阶段
    for (int i = 0; i < handle->food_number; i++) {
        new_food = alg_vector_from_matrix_row(handle->employed_bees, i);
        search_new_food(handle, new_food);
        new_fitness = handle->evaluate(new_food);
        if (new_fitness < handle->fitness->vector[i]) {
            // 更新食物源和适应度
            alg_vector_set_val(handle->fitness, i, new_fitness);
            for (int iter = 0; iter < handle->dim; iter++) {
                alg_matrix_set_val(handle->food_matrix, i, iter, *alg_vector_get_val(new_food, iter));
            }
        }
        alg_vector_free(new_food);
    }

    // 2. 观察者蜜蜂搜索阶段
    sum = vector_sum(handle->fitness); // 计算适应度的总和
    for (int i = 0; i < handle->food_number; i++) {
        prob = alg_safe_divide(handle->fitness->vector[i], sum); // 计算选择该食物源的概率
        if (isnan(prob)) {
            ERROR("GET A NAN VAL");
            continue;
        }

        rand = alg_random_float64(0, 1); // 生成一个 0 到 1 之间的随机数
        if (rand > prob) {
            // 随机数大于概率值，不选择该食物源，跳过
            continue;
        }

        // 选择该食物源进行局部搜索
        new_food = alg_vector_from_matrix_row(handle->onlooker_bees, i);
        search_new_food(handle, new_food);
        new_fitness = handle->evaluate(new_food);
        if (new_fitness < handle->fitness->vector[i]) {
            // 更新食物源和适应度
            alg_vector_set_val(handle->fitness, i, new_fitness);
            for (int iter = 0; iter < handle->dim; iter++) {
                alg_matrix_set_val(handle->food_matrix, i, iter, *alg_vector_get_val(new_food, iter));
            }
        }
        alg_vector_free(new_food);
    }

    return ALG_OK;
}

alg_state abc_free(abc_handle *handle) {
    alg_matrix_free(handle->food_matrix);
    alg_matrix_free(handle->employed_bees);
    alg_matrix_free(handle->onlooker_bees);
    alg_vector_free(handle->fitness);
    ALG_FREE(handle);
    return ALG_OK;
}